File: src/main/cpp/narang.prateek/recursion/print_numbers.cpp
#include<bits/stdc++.h>

using namespace std;

map<int, string> m{
    {0, "zero"},
    {1, "one"},
    {2, "two"},
    {3, "three"},
    {4, "four"},
    {5, "five"},
    {6, "six"},
    {7, "seven"},
    {8, "eight"},
    {9, "nine"},
};


void print_number(int n) {
    if (n / 10 == 0) {
        cout << m[n] << ' ';
        return;
    }
    print_number(n / 10);
    cout << m[n%10] << ' ';
}

int main() {
    print_number(2021);
}

File: src/main/java/helio/pedrini/recursividade/PrintNumbers.java
package helio.pedrini.recursividade;

import java.util.Map;

public class PrintNumbers {

    private static final Map<Integer, String> MAP = Map.of(
            0, "zero",
            1, "one",
            2, "two",
            3, "three",
            4, "four",
            5, "five",
            6, "six",
            7, "seven",
            8, "eight",
            9, "nine"
    );

    private static void printNumbers(int n) {
        if (n / 10 == 0) {
            System.out.printf("%s ", MAP.get(n));
            return;
        }
        printNumbers(n / 10);
        System.out.printf("%s ", MAP.get(n % 10));
    }

    public static void main(String[] args) {
        printNumbers(2021);
    }
}
